server/dex/dex.go

```diff
@@ -0,0 +1,254 @@
+		// asset symbol must be available.
+		subsys := fmt.Sprintf("ASSET[%s]", symbol)
+		logger := slog.NewBackend(os.Stdout).Logger(subsys)
+		log.Infof("Starting asset backend %q...", symbol)
+		be, err := asset.Start(symbol, ctxDex, assetConf.ConfigPath, logger, cfg.Network)
+		if err != nil {
+			cancel()
+			return nil, fmt.Errorf("failed to start asset %q: %v", symbol, err)
+		}
+
+		if symbol == "dcr" {
+			var ok bool
+			dcrBackend, ok = be.(*dcrasset.DCRBackend)
+			if !ok {
+				cancel()
+				return nil, fmt.Errorf("dcr backend is invalid")
+			}
+		}
+
+		ba := &asset.BackedAsset{
+			Asset: dex.Asset{
+				ID:       ID,
+				Symbol:   symbol,
+				LotSize:  assetConf.LotSize,
+				RateStep: assetConf.RateStep,
+				FeeRate:  assetConf.FeeRate,
+				SwapConf: assetConf.SwapConf,
+				FundConf: assetConf.FundConf,
+			},
+			Backend: be,
+		}
+
+		backedAssets[ID] = ba
+		lockableAssets[ID] = &swap.LockableAsset{
+			BackedAsset: ba,
+			CoinLocker:  dexCoinLocker.AssetLocker(ID).Swap(),
+		}
+	}
+
+	// Create DEXArchivist with the pg DB driver.
+	pgCfg := &pg.Config{
+		Host:          cfg.DBConf.Host,
+		Port:          strconv.Itoa(int(cfg.DBConf.Port)),
+		User:          cfg.DBConf.User,
+		Pass:          cfg.DBConf.Pass,
+		DBName:        cfg.DBConf.DBName,
+		HidePGConfig:  false,
+		QueryTimeout:  20 * time.Minute,
+		MarketCfg:     cfg.Markets,
+		CheckedStores: true,
+		Net:           cfg.Network,
+		FeeKey:        cfg.FeePubKey,
+	}
+	storage, err := db.Open(ctxDex, pgCfg)
+	if err != nil {
+		cancel()
+		return nil, fmt.Errorf("db.Open: %v", err)
+	}
+
+	authCfg := auth.Config{
+		Storage: storage,
+		Signer:  cfg.DEXPrivKey,
+		//StartEpoch // ! different for each asset, and ambiguous
+		RegistrationFee: cfg.RegFeeAmount,
+		FeeConfs:        cfg.RegFeeConfirms,
+		FeeChecker:      dcrBackend.UnspentCoinDetails,
+	}
+
+	authMgr := auth.NewAuthManager(&authCfg)
+
+	// Create the swapper.
+	swapperCfg := &swap.Config{
+		Ctx:              ctxDex,
+		Assets:           lockableAssets,
+		Storage:          storage,
+		AuthManager:      authMgr,
+		BroadcastTimeout: 10 * time.Second,
+	}
+
+	swapper := swap.NewSwapper(swapperCfg)
+
+	// Markets
+	markets := make(map[string]*market.Market, len(cfg.Markets))
+	for _, mktInf := range cfg.Markets {
+		baseCoinLocker := dexCoinLocker.AssetLocker(mktInf.Base).Book()
+		quoteCoinLocker := dexCoinLocker.AssetLocker(mktInf.Quote).Book()
+		mkt, err := market.NewMarket(ctxDex, mktInf, storage, swapper, authMgr, baseCoinLocker, quoteCoinLocker)
+		if err != nil {
+			cancel()
+			return nil, fmt.Errorf("NewMarket failed: %v", err)
+		}
+		markets[mktInf.Name] = mkt
+	}
+
+	// Book router
+	bookSources := make(map[string]market.BookSource, len(cfg.Markets))
+	for mktName, mkt := range markets {
+		bookSources[mktName] = mkt
+	}
+	bookRouter := market.NewBookRouter(&market.BookRouterConfig{
+		Ctx:     ctxDex,
+		Sources: bookSources,
+		//EpochDuration: 10_000, // !!!!! different for each market
```

Comment:
Could be added to Sources so the book router can get each market's epoch duration.

Reply:
Oh, but I put the epoch index into the order feed's message so that BookRouter doesn't have to do any time computations with epoch duration, which could be incorrect anyway.  Only the Market's epoch processing loop should be setting the order's time stamp and thus epoch index.